src/stdlib_experimental_quadrature_simps.fypp

```diff
@@ -0,0 +1,278 @@
+        ${t1}$, dimension(:), intent(in) :: y
+        ${t1}$, intent(in) :: dx
+        integer, intent(in), optional :: even
+        ${t1}$ :: integral
+
+        integer :: n
+
+        n = size(y)
+
+        select case (n)
+        case (0:1)
+            integral = 0.0_${k1}$
+        case (2)
+            integral = 0.5_${k1}$*dx*(y(1) + y(2))
+        case (3)
+            integral = dx/3.0_${k1}$*(y(1) + 4*y(2) + y(3))
+        case (4)
+            integral = simps38(y, dx)
+        ! case (5) not needed; handled by default
+        case (6) ! needs special handling because of averaged 3/8's rule case
+            if (present(even)) then
+                if (even < 0) then
+                    ! 3/8 rule on left
+                    integral =  simps38(y(1:4), dx) + simps(y(4:6), dx)
+                    return
+                else if (even > 0) then
+                    ! 3/8 rule on right
+                    integral = simps(y(1:3), dx) + simps38(y(3:6), dx)
+                    return
+                else
+                    ! fall through
+                end if
+            end if
+            ! either `even` not present or is zero
+            ! equivalent to averaging left and right
+            integral = dx/48.0_${k1}$ * (17*(y(1) + y(6)) + 59*(y(2) + y(5)) + 44*(y(3) + y(4)))
+        case default
+            if (mod(n, 2) == 1) then
+                integral = dx/3.0_${k1}$*(y(1) + 4*sum(y(2:n-1:2)) + 2*sum(y(3:n-2:2)) + y(n))
+            else
+                if (present(even)) then
+                    if (even < 0) then
+                        ! 3/8th rule on left
+                        integral = simps38(y(1:4), dx) + simps(y(4:n), dx)
+                        return
+                    else if (even > 0) then
+                        ! 3/8 rule on right
+                        integral = simps(y(1:n-3), dx) + simps38(y(n-3:n), dx)
+                        return
+                    else
+                        ! fall through
+                    end if
+                end if
+                ! either `even` not present or is zero
+                ! equivalent to averaging left and right
+                integral = dx/48.0_${k1}$ * (17*(y(1) + y(n)) + 59*(y(2) + y(n-1)) &
+                        + 43*(y(3) + y(n-2)) + 49*(y(4) + y(n-3)) + 48*sum(y(5:n-4)))
+            end if
+        end select
+    end function simps_dx_${k1}$
+
+#:endfor
+#:for k1, t1 in REAL_KINDS_TYPES
+
+    pure recursive module function simps_x_${k1}$(y, x, even) result(integral)
+        ${t1}$, dimension(:), intent(in) :: y
+        ${t1}$, dimension(:), intent(in) :: x
+        integer, intent(in), optional :: even
+        ${t1}$ :: integral
+
+        integer :: i
+        integer :: n
+
+        ${t1}$ :: h1, h2
+        ${t1}$ :: a, b, c
+
+        n = size(y)
+        if (size(x) /= n) error stop "simps: Arguments `x` and `y` must be the same size."
```

Comment:
I suggest you use the `stdlib` subroutine `error_stop` to avoid issues with older Fortran standards

```suggestion
        if (size(x) /= n) call error_stop("simps: Arguments `x` and `y` must be the same size.")
```
The functions cannot be pure when calling error_stop

Reply:
I think this is a defect of our `error_stop`. The only functionality it adds over a bare `error stop` statement is to optionally put both a message and an error code. Because that message is written to stderr, purity is lost. At least with a bare `error stop`, we can keep purity with sufficiently recent compilers (e.g., GCC version 8 or newer). I am perfectly happy to pretend pre-f2018 standards do not exist, but I think that does not align with stdlib's current goals. 

The alternative is to revert to declaring `x` to be `dimension(size(y))`. This has the potential performance penalties described in other comments, but at least the compiler/runtime will do the error checking and reporting for us without running afoul of purity rules.

Reply:
I don't think reverting to declaring `x` to be `dimension(size(y))` is a good idea, as mentioned in earlier comments.

It has been agreed that Gfortran <=7 were not supported for stdlib. However, recent versions of other compilers may not support the 2018 `error stop` yet. I am perfectly ok to use the 2018 `error stop`, but this will restrict stdlib to very recent compilers.